### PR TITLE
Distillation function

### PR DESCRIPTION
**Pull request by colbybanbury**

### CODE REVIEW

File: distill.py [hunk 1]
```diff
@@ -0,0 +1,218 @@
...
+
+from experiment_config import default_cfg, get_cfg
+from wake_vision_loader import get_wake_vision, get_miaps
+from vww_loader import get_vww
+
+import wandb
+from wandb.keras import WandbMetricsLogger
+
+
+def distill(teacher_config, student_cfg=default_cfg, extra_evals=["distance_eval", "miap_eval"]):
```
Comment (Ekhao) [inline distill.py:26]: Is there any reason that lighting is not included as an extra eval?

File: distill.py [hunk 2]
```diff
@@ -0,0 +1,218 @@
...
+    wandb.init(project="wake-vision", name=student_cfg.EXPERIMENT_NAME+"_Distill", config=student_cfg)
```
Comment (Ekhao) [inline distill.py:27]:
```suggestion
    wandb.init(entity="harvard-edge", project="wake-vision", name=student_cfg.EXPERIMENT_NAME+"_Distill", config=student_cfg)
```

File: distill.py [hunk 3]
```diff
@@ -0,0 +1,218 @@
...
+            Args:
+                optimizer: Keras optimizer for the student weights
+                metrics: Keras metrics for evaluation
+                student_loss_fn: Loss function of difference between student
+                    predictions and ground-truth
+                distillation_loss_fn: Loss function of difference between soft
+                    student predictions and soft teacher predictions
+                alpha: weight to student_loss_fn and 1-alpha to distillation_loss_fn
+                temperature: Temperature for softening probability distributions.
+                    Larger temperature gives softer distributions.
```
Comment (Ekhao) [inline distill.py:81]: It seems that these arg descriptions are not completely corresponding to the function arguments

File: distill.py [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+            """
+            super().compile(optimizer=optimizer, metrics=metrics, **kwargs)
+            self.student_loss_fn = student_loss_fn
+            self.distillation_loss_fn = distillation_loss_fn
+            self.alpha = alpha
+
+        def compute_loss(
+            self, x=None, y=None, y_pred=None, sample_weight=None, allow_empty=False
```
Comment (Ekhao) [inline distill.py:89]: Why should x and y be None by default?

Comment (Ekhao) [reply distill.py:89]: Same with y_pred, and also where is sample_weight and allow_empty used?

File: distill.py [hunk 5]
```diff
@@ -0,0 +1,218 @@
...
+    #         kl_loss = []
+    #         for batch in val:
+    #             images, _ = batch
+    #             teacher_pred = teacher(images, training=False)
+    #             student_pred = self.model(images, training=False)
+    #             kl_loss.append(keras.losses.KLDivergence()(teacher_pred, student_pred))
+    #         kl_loss = np.mean(kl_loss)
+    #         print(f"KL-Divergence: {kl_loss}")
+    #         wandb.log({"epoch/KL-Divergence": kl_loss})
+    # callbacks.append(distillCallback())
```
Comment (Ekhao) [inline distill.py:165]: Is this supposed to be deleted?

File: distill.py (hunk 1 above)
Comment (colbybanbury) [reply distill.py]: No that was just how it was set up originally. 

I'm actually going to remove validation evals for now since it slows things down and they aren't f1. 

We can add them back to distill later but for now we mostly care about the test set versions for finer grain eval.
